crates/velyst_macros/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+use darling::{
+    ast::Data,
+    util::{Flag, Ignored},
+    Error, FromDeriveInput, FromField,
+};
+use proc_macro::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{parse_macro_input, spanned::Spanned, DeriveInput, Ident, Index};
+
+#[derive(FromDeriveInput)]
+#[darling(forward_attrs(typst_path))]
+struct TypstPathArgs {
+    ident: Ident,
+    #[darling(with = "convert_typst_path")]
+    attrs: TypstPathValue,
+}
+
+struct TypstPathValue(syn::Expr);
+
+fn convert_typst_path(attrs: Vec<syn::Attribute>) -> darling::Result<TypstPathValue> {
+    let mut filtered = attrs
+        .into_iter()
+        .filter(|a| a.path().is_ident("typst_path"));
+
+    let attr = filtered
+        .next()
+        .ok_or(darling::Error::custom("#[typst_path] is required"))?;
+
+    if filtered.next().is_some() {
+        return Err(darling::Error::custom("#[typst_path] can only occur once"));
+    }
+
+    let val = attr.meta.require_name_value()?.value.clone();
+
+    Ok(TypstPathValue(val))
+}
+
+/// Macro for deriving `TypstPath`.
+///
+/// Usage
+/// ```rust
+/// # use velyst_macros::TypstPath;
+/// #[derive(TypstPath)]
+/// #[typst_path = "path_to_file.typ"] // Path relative to bevy asset dir
+/// struct MyTypstFile;
+/// ```
+#[proc_macro_derive(TypstPath, attributes(typst_path))]
+pub fn derive_typst_path(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let TypstPathArgs {
+        ident,
+        attrs: TypstPathValue(path),
+    } = match TypstPathArgs::from_derive_input(&ast) {
+        Ok(t) => t,
+        Err(e) => return e.write_errors().into(),
+    };
+
+    quote! {
+        impl velyst::renderer::TypstPath for #ident {
+            fn path() -> &'static str {
+                #path
+            }
+        }
+    }
+    .into()
+}
+
+#[derive(FromDeriveInput)]
+#[darling(supports(struct_any), attributes(typst_func))]
+struct TypstFuncArgs {
+    ident: Ident,
+    data: Data<Ignored, TypstFuncField>,
+    name: String,
+    #[darling(default)]
+    layer: usize,
+}
+
+#[derive(FromField)]
+#[darling(attributes(typst_func))]
+struct TypstFuncField {
+    ident: Option<Ident>,
+    named: Flag,
+}
+
+/// Macro for deriving `TypstFunc`.
+///
+/// Usage
+/// ```rust
+/// # use velyst_macros::TypstFunc;
+/// # use bevy::prelude::*;
+/// #[derive(Resource, TypstFunc)]
+/// #[typst_func(name = "main", layer = 0)] // layer is optional
+/// struct MainFunc {
+///     width: f64,
+///     height: f64,
+///     #[typst_func(named)] // use as a named argument
```

Comment:
is it better to just use `#[named]` ?

Also, we could add an optional named parameter as well: `#[named = "abc"]` to override the current field name (in case needed).

Reply:
As far as I can tell, it seems to be a convention in the rust ecosystem for each derive macro to only have one dedicated attribute, and to have all configuration nested inside this attribute. This especially makes sense when the user wants to derive other Traits as well, which introduce their own attributes. For one, it would be hard to tell which attribute belongs to which derive macro, and in the worst case there could be name collisions (since "named" is a pretty generic name).

The optional override is a good idea tho, i will add that later.

Reply:
Alright. Sounds good!